resolwe/flow/models/data.py

```diff
@@ -319,31 +322,58 @@ def create_entity(self):
         any `Entity`, create new `Entity`
 
         """
-        ds_slug = self.process.flow_collection  # pylint: disable=no-member
-        if ds_slug:
-            entity_query = Entity.objects.filter(data__in=self.parents.all()).distinct()  # pylint: disable=no-member
-
-            if entity_query.count() == 1:
-                entity = entity_query.first()
-
-                copy_permissions(entity, self)
+        entity_type = self.process.entity_type  # pylint: disable=no-member
+        entity_descriptor_schema = self.process.entity_descriptor_schema  # pylint: disable=no-member
+        entity_input = self.process.entity_input  # pylint: disable=no-member
+
+        if entity_type:
+
+            if entity_input:
+                input_id = dict_dot(self.input, entity_input, default=lambda: None)
+                if input_id is None:
+                    logger.warning("Skipping creation of entity due to missing input.")
+                    return
+                if isinstance(input_id, int):
+                    entity_query = Entity.objects.filter(data__pk=input_id)
+                elif isinstance(input_id, list):
+                    entity_query = Entity.objects.filter(data__pk__in=input_id)
+                else:
+                    raise ValueError("")
             else:
+                entity_query = Entity.objects.filter(
+                    data__in=self.parents.all(),  # pylint: disable=no-member
+                    type=entity_type,
+                )
 
-                descriptor_schema = DescriptorSchema.objects.filter(slug=ds_slug).latest()
+            entity_query = entity_query.distinct()
+            entity_count = entity_query.count()
+
+            if entity_count == 0:
+                descriptor_schema = DescriptorSchema.objects.filter(
+                    slug=entity_descriptor_schema
+                ).latest()
                 entity = Entity.objects.create(
                     contributor=self.contributor,
                     descriptor_schema=descriptor_schema,
+                    type=entity_type,
                     name=self.name,
                     tags=self.tags,
                 )
-
                 assign_contributor_permissions(entity)
 
-            entity.data.add(self)
+            elif entity_count == 1:
+                entity = entity_query.first()
+                copy_permissions(entity, self)
 
-            # Inherite collections from entity.
-            for collection in entity.collections.all():
-                collection.data.add(self)
+            else:
+                logger.info("Skipping creation of entity due to multiple entities found.")
+                entity = None
+
+            if entity:
+                entity.data.add(self)
```

Comment:
Maybe all this (entity creation, attaching data to entity and collections...) should be done in transaction?

Reply:
Or is the whole `save()` method called in transaction? Could you write a test?